### PR TITLE
Bug #54 fix: Enable CORS Support Amol

### PR DESCRIPTION
**Pull request by thite-amol**
This change is [Reviewable](https://reviewable.io/reviews/techjoomla/com_api/58)

### CODE REVIEW

File: code/admin/config.xml [hunk 1]
```diff
@@ -1,35 +1,42 @@
...
+			description="COM_API_FORM_LBL_RATE_LIMIT_DESC" />
+		<field name="log_requests" type="radio" default="0"
+			label="COM_API_CONFIG_LOG_LBL" description="COM_API_CONFIG_LOG_DESC"
+			class="btn-group">
+			<option value="0">No</option>
+			<option value="1">Yes</option>
+		</field>
+		<field name="exclude_log" type="text" default="password,pass,passwd,pwd,key"
+			label="COM_API_EXCLD_WORDS" description="COM_API_EXCLD_WORDS_DESC" />
+		<field name="exclude_log" type="text" default="password,pass,passwd,pwd,key"
```
Comment (coolbung) [inline code/admin/config.xml]: Why is this field there twice ?

File: code/admin/language/en-GB/en-GB.com_api.ini [hunk 2]
```diff
@@ -90,6 +90,9 @@ COM_API_FORM_LBL_RATE_LIMIT_DESC="Put a number if you want to limit the number o
 COM_API_EXCLD_WORDS="Exclude request variables from log"
 COM_API_EXCLD_WORDS_DESC="A comma separated list of request variables that will be redacted before being added to the API Request log"
 COM_API_FORM_DESC_FORCE_OUTPUT="Force output params"
+COM_API_CONFIG_ALLOW_HEADER_LBL="Values for Access-Control-Allow-Headers"
+COM_API_CONFIG_ALLOW_HEADER_DESC=" Add comma seperated values for Access-Control-Allow-Headers"
```
Comment (coolbung) [inline code/admin/language/en-GB/en-GB.com_api.ini]: Remove the space before sentence

File: code/site/controllers/http.php [hunk 3]
```diff
@@ -40,22 +40,35 @@ public function display($cachable = false, $urlparams = array())
...
-			header("Access-Control-Allow-Origin: " . $cors_urls);
+			$corsUrls = $params->get('cors', "*");
+
+			if ($corsUrls === "*")
+			{
+				header("Access-Control-Allow-Origin: " . '*');
+			}
+			else
+			{
+				$httpOrigin = $app->input->server->getString('HTTP_REFERER', '');
```
Comment (coolbung) [inline code/site/controllers/http.php]: The referer may give a full URL like http://truck.com/jobs and not just http://truck.com 

Is this being taken care of ?

File: code/site/controllers/http.php [hunk 4]
```diff
@@ -40,22 +40,35 @@ public function display($cachable = false, $urlparams = array())
...
+				}
+			}
+
 			header("Access-Control-Allow-Methods: " . strtoupper($params->get("allow_cors")));
 		}
 
 		if ($callMethod === "OPTIONS")
 		{
-			header("Content-type: application/json");
+			header("Content-type: text/html; charset=utf-8");
```
Comment (coolbung) [inline code/site/controllers/http.php]: Is this the right content type for OPTIONS call ?

File: code/site/controllers/http.php [hunk 5]
```diff
@@ -40,22 +40,38 @@ public function display($cachable = false, $urlparams = array())
...
 		$params = JComponentHelper::getParams('com_api');
+		$callMethod = $app->input->getMethod();
+		$httpOrigin = $app->input->server->getString('HTTP_REFERER', '');
 
-		// Set CORS header add comma seperated values directaly in input box
-		$cors_urls = $params->get('cors', "*");
-
-		$callMethod = $app->input->server->get('REQUEST_METHOD', '', 'STRING');
+		$JUriObj = JUri::getInstance($httpOrigin);
+		$referer = $JUriObj->toString($parts = array('scheme', 'host'));
```
Comment (coolbung) [inline code/site/controllers/http.php]: Is the `$parts` variable needed here ?

File: code/site/controllers/http.php [hunk 6]
```diff
@@ -40,22 +40,38 @@ public function display($cachable = false, $urlparams = array())
...
-			header("Access-Control-Allow-Origin: " . $cors_urls);
+			$corsUrls = $params->get('cors', "*");
+
+			if ($corsUrls === "*")
+			{
+				header("Access-Control-Allow-Origin: " . '*');
+			}
+			else
+			{
+				$corsUrlsArray = array_map('trim', array_filter(explode(',', $corsUrls)));
```
Comment (coolbung) [inline code/site/controllers/http.php]: What's the purpose of using array_filter here ? Can we simplify the code.

**Comment by sushan05** [comment]
@thite-amol @coolbung could we push this MR to be merged as early as possible if it resolves the fundamental issue of CORS.

**Comment by coolbung** [comment]
Can one of you address the comments that I have added. 

Also take a look at the other issues reported by the static code analysers.

**Pull request merged**